### DISCUSSION TITLE
Load custom TTF font from local filesystem

### DISCUSSION THREAD

**Discussion by aprat84** [dompdf/dompdf#3127] [Q&A] [answered]
Hi!

_I've been reading through pracically all issues and discussion posts relative to loading fonts, with no luck!_

I'm trying to load custom TTF fonts, precisely, Source Sans Pro downloaded from Google Fonts site.
Loading external fonts via HTTP is not an option (it's gross and we don't want to rely on connections to third party servers).

dompdf version: `2.0.2`

Symfony project in `/var/www/html`

Assets directory:
<img width="345" alt="image" src="https://user-images.githubusercontent.com/39481795/216587624-a9ae1335-eaf0-4a80-b673-3faa90f9ca3e.png">

Markup:

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Test</title>
    <style>
        @font-face {
            font-family: 'Source Sans Pro';
            font-style: normal;
            font-weight: 400;
            src: url('fonts/source-sans-pro/SourceSansPro-Regular.ttf') format('truetype');
        }
        /* Other @font-face declarations with different weight/style */ 

        html, body {
            color: #303030;
            font-family: "Source Sans Pro", serif;
            font-size: 9pt;
        }
    </style>
</head>
<body>
    <p>Test PDF!</p>
</body>
</html>
```

PHP code:
```php
$html = "..."; // Twig render

$pdf = new Dompdf([
    'chroot'    => $this->parameterBag->get('kernel.project_dir').'/assets',
    'fontDir'   => $this->parameterBag->get('kernel.cache_dir').'/dompdf/fonts',
    'fontCache' => $this->parameterBag->get('kernel.cache_dir').'/dompdf/fonts',
]);
$pdf->setPaper('A4');
$pdf->loadHtml($html);
$pdf->render();

return $pdf->output();
```

From my point of view, the `@font-face` url is a correct relative path inside the `chroot` path configured to dompdf, I don't know why it doesn't load the fon... I think I've tried everything possible...

The worst part is that there's no error or warning, it simply gets ignored by this line: https://github.com/dompdf/dompdf/blob/2a8a6b80fcaa5148ace50f35a10979fe00c6a35d/src/Css/Stylesheet.php#L1459

**Comment by bsweeney** [comment]
Since you've peeked at what's happening in the source have you confirmed that the URI in $source["path"] is correct?

You don't indicate where the PHP file is located in your file path. Since you're using loadHTML the base path will be the location of that file. Based on the HTML, your PHP file would need to be located in the assets directory.

It's true that messaging is can be non-optimal since we haven't put in the time to build out PSR-3 support. However, in the meantime you can try to get additional information by [enabling additional debug messaging](https://github.com/dompdf/dompdf/wiki/Debugging-Dompdf).

**Comment by aprat84** [comment, marked as answer]
`$source['path']` is `null`, because the `Helpers::build_url()` can't resolve the font-face url (it's a relative path from chroot): https://github.com/dompdf/dompdf/blob/2a8a6b80fcaa5148ace50f35a10979fe00c6a35d/src/Helpers.php#L115

The PHP file is a classe in the `/src` directory namespace hierarchy.

Should I `$pdf->setBasePath($pdf->getOptions()->getChroot());` ?
Edit: I've tested it and it works!

**Reply by bsweeney** [reply]
There are a few ways you could achieve this:
- setting the base path
- setting a base href in the HTML document
- move the PHP so that relative path is correct from that file

I think the way you went about it is perhaps the best option. Note that things get trickier if you're using absolute paths (starting with /) since those would be read from the root of the filesystem. Luckily that's not the case here.

**Reply by aprat84** [reply]
Yes, that's what I thought:
* setting the base path in HTML is dirty, cause it is an absolute path from the filesystem
* moving the PHP file is not possible, and it won't make sense in a symfony project to have a PHP file inside assets directory

Thank you!
